### PR TITLE
feat(core): Implement RichString core data structures and Span overlap resolution

### PR DESCRIPTION
**Pull request by mkeeda**
## Description
This PR introduces the core data structures for the RichText library, implementing a type-safe, immutable foundation to manage formatted text ranges for Jetpack Compose.

## Key Features & Implementations

1. **`AttributeContainer` & `AttributeKey`**
   - Implemented a type-safe, immutable attribute map configuration.
   - Provided an elegant `merge` (`+` operator) allowing for easy cascading of attribute priorities.

2. **`RichString` & `RichSpan`**
   - Core immutable rich text representations that pair plain text with attributes applied to specific character `IntRange`s.
   - Operations like `with(key, value, range)` always return a completely new instance, ensuring immutability across the library.

3. **Strict Range Boundaries Validation**
   - Prevents invalid state creations by ensuring robust validations.
   - Range configurations securely reject out-of-bounds queries, negative numbers, or mathematically invalid empty ranges prior to spans logic.

4. **Automatic Overlap Resolution (Sweep-line Algorithm)**
   - The highlight feature implemented in `SpanMerger`.
   - Instead of maintaining dirty, loosely stacked spans, the library incorporates a sweep-line style interval chunking algorithm. 
   - When placing an overlapping attribute over existing spans, this algorithm acts with `O(N)` precision to automatically divide, merge, and optimize interval bounds. It uniquely guarantees the internal span list natively remains strictly **non-overlapping** and sorted at all times.

## Verification & Testing
- Integrated strict TDD lifecycle utilizing JUnit 4 & Kotest properties.
- **Assertions covered:** 
  - Edge cases on boundaries exceptions.
  - Partial overlaps scaling and dividing.
  - Totally engulfed or "contained" sub-spans interactions.
  - Performance-optimized back-to-back merging of identical contiguous attributes.

### CODE REVIEW

File: core/src/main/kotlin/dev/mkeeda/arranger/core/text/AttributeContainer.kt [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    /**
+     * Returns a new [AttributeContainer] containing the specified key and value.
+     * If a key with the same name already exists, it is overwritten with the new value.
+     */
+    public fun <T> with(key: RichAttributeKey<T>, value: T): AttributeContainer {
+        val newAttributes = attributes.toMutableMap()
+        newAttributes[key] = value
+        return AttributeContainer(newAttributes)
+    }
```
Comment (mkeeda) [inline core/src/main/kotlin/dev/mkeeda/arranger/core/text/AttributeContainer.kt]: Kotlinの `+` 演算子を使うことで、ミュータブルなコレクションを隠蔽しつつより関数型らしく簡潔に書くことができます。また、プロジェクトの引数渡し規約（ラベルの使用）に従いやすくなるため、以下のような書き方を提案します。

```kotlin
public fun <T> with(key: RichAttributeKey<T>, value: T): AttributeContainer {
    return AttributeContainer(attributes = attributes + (key to value))
}
```

File: core/src/main/kotlin/dev/mkeeda/arranger/core/text/AttributeContainer.kt [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+     * Returns a new [AttributeContainer] containing all attributes from this and the [other].
+     * If keys collide, values from [other] take precedence (overwrite).
+     */
+    public operator fun plus(other: AttributeContainer): AttributeContainer {
+        if (other.attributes.isEmpty()) return this
+        if (this.attributes.isEmpty()) return other
+        val newAttributes = attributes.toMutableMap()
+        newAttributes.putAll(other.attributes)
+        return AttributeContainer(newAttributes)
+    }
```
Comment (mkeeda) [inline core/src/main/kotlin/dev/mkeeda/arranger/core/text/AttributeContainer.kt:49]: 上記同様に、Map の `+` オペレーターを利用したリファクタリングの提案です。

```kotlin
public operator fun plus(other: AttributeContainer): AttributeContainer {
    if (other.attributes.isEmpty()) return this
    if (this.attributes.isEmpty()) return other
    return AttributeContainer(attributes = attributes + other.attributes)
}
```

File: core/src/main/kotlin/dev/mkeeda/arranger/core/text/AttributeContainer.kt [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        if (javaClass != other?.javaClass) return false
+
+        other as AttributeContainer
+
+        return attributes == other.attributes
+    }
+
+    override fun hashCode(): Int {
+        return attributes.hashCode()
+    }
```
Comment (mkeeda) [inline core/src/main/kotlin/dev/mkeeda/arranger/core/text/AttributeContainer.kt:62]: Nitpick: コードスタイル規約（ラベル付き引数）へ準拠するため、`attributes =` というラベルの追加を提案します。

```kotlin
private val EMPTY = AttributeContainer(attributes = emptyMap())
```

File: core/src/main/kotlin/dev/mkeeda/arranger/core/text/SpanMerger.kt [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    // 1. Extract boundaries and sort uniquely
+    val boundaries = mutableListOf<Int>()
+    boundaries.add(newSpan.range.first)
+    boundaries.add(newSpan.range.last + 1)
+
+    for (span in existingSpans) {
+        boundaries.add(span.range.first)
+        boundaries.add(span.range.last + 1)
+    }
```
Comment (mkeeda) [inline core/src/main/kotlin/dev/mkeeda/arranger/core/text/SpanMerger.kt]: Nitpick: `mutableListOf` と `add` の連続呼び出しでも機能的には全く問題ありませんが、標準関数の `buildList` を使うとスコープが閉じてよりKotlinのIdiomらしくなります。

```kotlin
val boundaries = buildList {
    add(newSpan.range.first)
    add(newSpan.range.last + 1)
    for (span in existingSpans) {
        add(span.range.first)
        add(span.range.last + 1)
    }
}
```

**Review by mkeeda — COMMENTED** [review]
The overall structure and algorithms (especially the Sweep-line logic in `SpanMerger`) are extremely robust and well-designed! The test coverage is also comprehensive, effectively handling various edge cases and boundaries. This forms a fantastic foundational layer for the RichString representations.

I left a few inline comments with suggestions to make the code even more idiomatic to Kotlin and to adhere closely to the project's parameter label guidelines.

Great work! LGTM 🚀

File: core/src/main/kotlin/dev/mkeeda/arranger/core/text/AttributeContainer.kt (hunk 1 above)
Comment (mkeeda) [reply core/src/main/kotlin/dev/mkeeda/arranger/core/text/AttributeContainer.kt]: *(AI review) English version:*
By using Kotlin's `+` operator, we can conceal the mutable collection and write this in a more concise, functional manner. Additionally, it makes it easier to adhere to the project's parameter passing guidelines (using labels), so I suggest the following approach:

```kotlin
public fun <T> with(key: RichAttributeKey<T>, value: T): AttributeContainer {
    return AttributeContainer(attributes = attributes + (key to value))
}
```

File: core/src/main/kotlin/dev/mkeeda/arranger/core/text/AttributeContainer.kt (hunk 2 above)
Comment (mkeeda) [reply core/src/main/kotlin/dev/mkeeda/arranger/core/text/AttributeContainer.kt:49]: *(AI review) English version:*
Similarly, here's a suggestion for refactoring using the Map's `+` operator.

```kotlin
public operator fun plus(other: AttributeContainer): AttributeContainer {
    if (other.attributes.isEmpty()) return this
    if (this.attributes.isEmpty()) return other
    return AttributeContainer(attributes = attributes + other.attributes)
}
```

File: core/src/main/kotlin/dev/mkeeda/arranger/core/text/AttributeContainer.kt (hunk 3 above)
Comment (mkeeda) [reply core/src/main/kotlin/dev/mkeeda/arranger/core/text/AttributeContainer.kt:62]: *(AI review) English version:*
Nitpick: To comply with the code style guide (labeled arguments), I suggest adding the `attributes =` label.

```kotlin
private val EMPTY = AttributeContainer(attributes = emptyMap())
```

File: core/src/main/kotlin/dev/mkeeda/arranger/core/text/SpanMerger.kt (hunk 4 above)
Comment (mkeeda) [reply core/src/main/kotlin/dev/mkeeda/arranger/core/text/SpanMerger.kt]: *(AI review) English version:*
Nitpick: While calling `mutableListOf` and `add` sequentially is perfectly fine functionally, using the standard `buildList` function closes the scope and feels much more idiomatic to Kotlin.

```kotlin
val boundaries = buildList {
    add(newSpan.range.first)
    add(newSpan.range.last + 1)
    for (span in existingSpans) {
        add(span.range.first)
        add(span.range.last + 1)
    }
}
```
